### ISSUE TITLE
Unfold measures when using PLE

### ISSUE THREAD

**Issue by facundominguez** [ucsd-progsys/liquid-fixpoint#489]
It will take some debugging to figure out why this workaround was needed in liquidhaskell:
https://github.com/ucsd-progsys/liquidhaskell/pull/1874

I'm guessing that liquid-fixpoint should unfold measures with PLE, but please comment if we should consider other alternatives.

**Comment by facundominguez** [comment]
The following change undoes the effect of 1874 on `tests/pos/T1874.hs`
```
diff --git a/src/Language/Haskell/Liquid/Bare.hs b/src/Language/Haskell/Liquid/Bare.hs
index 38e14c564..3d2a0c016 100644
--- a/src/Language/Haskell/Liquid/Bare.hs
+++ b/src/Language/Haskell/Liquid/Bare.hs
@@ -396,7 +396,7 @@ makeLiftedSpec0 :: Config -> GhcSrc -> F.TCEmb Ghc.TyCon -> LogicMap -> Ms.BareS
                 -> Ms.BareSpec
 makeLiftedSpec0 cfg src embs lmap mySpec = mempty
   { Ms.ealiases  = lmapEAlias . snd <$> Bare.makeHaskellInlines (typeclass cfg) src embs lmap mySpec 
-  , Ms.reflects  = Ms.reflects mySpec <> (if reflection cfg then Ms.hmeas mySpec else mempty)
+  , Ms.reflects  = Ms.reflects mySpec -- <> (if reflection cfg then Ms.hmeas mySpec else mempty)
   , Ms.dataDecls = Bare.makeHaskellDataDecls cfg name mySpec tcs  
   , Ms.embeds    = Ms.embeds mySpec
   -- We do want 'embeds' to survive and to be present into the final 'LiftedSpec'. The
```
This has LH fail with
```
**** LIQUID: UNSAFE ************************************************************

T1874.hs:68:15: error:
    Liquid Type Mismatch
    .
    The inferred type
      VV : T1874.Exp
    .
    is not a subtype of the required type
      VV : {VV : T1874.Exp | T1874.exprType e0 == T1874.exprType VV}
    .
    in the context
      e0 : T1874.Exp
    Constraint id 21
   |
68 |               Just e11 -- ? typeAppLam_prop ty e11 e2
   |               ^^^^^^^^
```
when I look at the equations produced by PLE in the constraint 21 I get surprisingly:
```
$ cat tests/pos/.liquid/T1874.hs.fq.ple
...
constraint id 21:
  && [((T1874.isFunTy T1874.TInt) = false);
      ((is$36$GHC.Types.$58$ GHC.Types.$91$$93$) = false);
      ((is$36$GHC.Types.$91$$93$ GHC.Types.$91$$93$) = true);
      ((is$T1874.TFun T1874.TInt) = false);
      ((is$T1874.TInt T1874.TInt) = true);
      ((isJust GHC.Maybe.Nothing) = false);
      ((isJust lq_anf$##7205759403792803251##d2vF) = false);
      ((isJust lq_anf$##7205759403792803253##d2vH) = false);
      ((len GHC.Types.$91$$93$) = 0)]
...
```
Which are too little equations, in particular, there is no sign of the measure function `exprType`, though other equalities from measures are there.

If we instead run the test with the unmodified LH, the module is safe and more equations are output:
```
$ cat tests/pos/.liquid/T1874.hs.fq.ple
...
constraint id 21:
  && [((T1874.exprType VV##F##21) =
         (if (is$T1874.Var VV##F##21) then
            (lqdc##$select##T1874.Var##2 VV##F##21)
          else
            (if (is$T1874.Lam VV##F##21) then
               (T1874.TFun (lqdc##$select##T1874.Lam##2 VV##F##21) (T1874.exprType (lqdc##$select##T1874.Lam##1 VV##F##21)))
             else
               (T1874.funResTy (T1874.exprType (T1874.e1 VV##F##21))))));
      ((T1874.exprType e0##aCL) =
         (T1874.funResTy (T1874.exprType (T1874.e1 e0##aCL))));
      ((T1874.exprType e0##aCL) =
         (if (is$T1874.Var (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL))) then
            (lqdc##$select##T1874.Var##2 (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL)))
          else
            (if (is$T1874.Lam (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL))) then
               (T1874.TFun (lqdc##$select##T1874.Lam##2 (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL))) (T1874.exprType (lqdc##$select##T1874.Lam##1 (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL)))))
             else
               (T1874.funResTy (T1874.exprType (T1874.e1 (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL))))))));
      ((T1874.exprType e1##aCM) =
         (T1874.TFun (lqdc##$select##T1874.Lam##2 e1##aCM) (T1874.exprType (lqdc##$select##T1874.Lam##1 e1##aCM))));
    ...
...
```

The fix in ucsd-progsys/liquidhaskell#1874 is putting `exprType` both as a measure and as a reflected function, but even if it were only a measure, PLE should pick it up.

cc @nikivazou

**Comment by facundominguez** [comment]
Looking at the environment of constraint 21 doesn't give me much of a clue yet. The data constructors `Lam` and `App` appear in the formulas, which should cause PLE to introduce `exprType` equations for them when it is only a measure.
```
constraint:
  lhs {VV : T1874.Exp | [$k_##925[VV##1744:=VV][VV##924:=VV][VV##F:=VV][lq_tmp$x##1745:=VV][lq_tmp$x##921:=VV##1741][lq_tmp$x##923:=e11][lq_tmp$x##926:=VV]]}
  rhs {VV : T1874.Exp | [((T1874.exprType e0) =
                            (T1874.exprType VV))]}
  id 21 tag [5]
  // META constraint id 21 : T1874.hs:68:15-22
  environment:
    // elided some likely irrelevant bindings
    e0 : {VV : T1874.Exp | []}
    
    e1 :
      {lq_tmp$x : T1874.Exp |
         [(T1874.isFunTy (T1874.exprType lq_tmp$x))]}
    
    e11 : {lq_tmp$x : T1874.Exp | []}
    
    e2 :
      {lq_tmp$x : T1874.Exp |
         [((T1874.funArgTy (T1874.exprType e1)) =
             (T1874.exprType lq_tmp$x))]}
    
    ty : {lq_tmp$x : T1874.Ty | []}

    _lq_anf$##d2vD :
      {lq_tmp$x : T1874.Exp |
         [(is$T1874.App lq_tmp$x);
          (~ ((is$T1874.Lam lq_tmp$x)));
          (~ ((is$T1874.Var lq_tmp$x)));
          (lq_tmp$x = e0);
          (lq_tmp$x = (T1874.App e1 e2));
          ((T1874.e1 lq_tmp$x) = e1);
          ((T1874.exprType lq_tmp$x) = (T1874.funResTy (T1874.exprType e1)));
          ((lqdc##$select##T1874.App##1 lq_tmp$x) = e2)]}
    
    _lq_anf$##d2vE :
      {lq_tmp$x : (GHC.Maybe.Maybe T1874.Exp) |
         [(lq_tmp$x = (T1874.step e1))]}
    
    _lq_anf$##d2vF :
      {lq_tmp$x : (GHC.Maybe.Maybe T1874.Exp) |
         [(~ ((isJust lq_tmp$x)));
          (lq_tmp$x = GHC.Maybe.Nothing);
          (lq_tmp$x = (T1874.step e1))]}
    
    _lq_anf$##d2vG :
      {lq_tmp$x : (GHC.Maybe.Maybe T1874.Exp) |
         [(lq_tmp$x = (T1874.step e2))]}
    
    _lq_anf$##d2vH :
      {lq_tmp$x : (GHC.Maybe.Maybe T1874.Exp) |
         [(~ ((isJust lq_tmp$x)));
          (lq_tmp$x = GHC.Maybe.Nothing);
          (lq_tmp$x = (T1874.step e2))]}
    
    _lq_anf$##d2vI :
      {lq_tmp$x : T1874.Exp |
         [(T1874.isFunTy (T1874.exprType lq_tmp$x));
          (is$T1874.Lam lq_tmp$x);
          (~ ((is$T1874.App lq_tmp$x)));
          (~ ((is$T1874.Var lq_tmp$x)));
          (lq_tmp$x = e1);
          (lq_tmp$x = (T1874.Lam ty e11));
          ((T1874.exprType lq_tmp$x) = (T1874.TFun ty (T1874.exprType e11)));
          ((lqdc##$select##T1874.Lam##1 lq_tmp$x) = e11);
          ((lqdc##$select##T1874.Lam##2 lq_tmp$x) = ty)]}
```

**Comment by ranjitjhala** [comment]
Interesting @nikivazou any ideas?

**Comment by facundominguez** [comment]
I have a fix!

The issue would be that in order for PLE to consider a subexpression as a "candidade" for analysis, according to the function `makeCandidates`, it needs to be `reducible` or an `if-then-else`, no condition of which is met by the occurrences of   our data constructors here.

The following fix gets verification going, we would have to check that it doesn't introduce performance regressions.

```
diff --git a/src/Language/Fixpoint/Solver/PLE.hs b/src/Language/Fixpoint/Solver/PLE.hs
index 167d295f..1ac0379b 100644
--- a/src/Language/Fixpoint/Solver/PLE.hs
+++ b/src/Language/Fixpoint/Solver/PLE.hs
@@ -401,7 +401,14 @@ makeCandidates :: Knowledge -> ICtx -> Expr -> [Expr]
 makeCandidates γ ctx expr 
   = mytracepp ("\n" ++ show (length cands) ++ " New Candidates") cands
   where 
-    cands = filter (\e -> isRedex γ e && not (e `S.member` icSolved ctx)) (notGuardedApps expr)
+    cands = filter (\e -> (isConstructorApplication γ e || isRedex γ e) && not (e `S.member` icSolved ctx)) (notGuardedApps expr)
+
+isConstructorApplication :: Knowledge -> Expr -> Bool
+isConstructorApplication γ e
+  | (EVar f, es) <- splitEApp e
+  = S.member f (knDCs γ)
+  | otherwise
+  = False
 
 isRedex :: Knowledge -> Expr -> Bool 
 isRedex γ e = isGoodApp γ e || isIte e 
```

**Comment by ranjitjhala** [comment]
Fantastic!!!

**Comment by nikivazou** [comment]
That's great! So, we should use this approach instead and revert my insufficient fix (ucsd-progsys/liquidhaskell#1874) that based on @michaelborkowski  slowed down performance!
